pkg/rekor/identity.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rekor
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+
+	"github.com/go-openapi/runtime"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/types"
+
+	hashedrekord_v001 "github.com/sigstore/rekor/pkg/types/hashedrekord/v0.0.1"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+// IdentityEntry holds a certificate subject, issuer, and log entry metadata
+type IdentityEntry struct {
+	Subject string
+	Issuer  string
+	Index   int64
+	UUID    string
+}
+
+// Identities are the subjects/issuers to search
+type Identities struct {
+	Identities []Identity `json:"identities,omitempty"`
+}
+
+type Identity struct {
+	Subject string   `json:"subject"`
+	Issuers []string `json:"issuers,omitempty"`
+}
+
+// MatchedIndices returns a list of log indices that contain the requested identities.
+func MatchedIndices(logEntries []models.LogEntry, identities Identities) ([]IdentityEntry, error) {
+	if err := verifyIdentities(identities); err != nil {
```

Comment:
shouldn't this check occur once at input ingestion?

Reply:
Hmmm agree, but on the other hand if the identities are badly formed as inputs, you probably don't want to do all the STH work before alerting on it. But not a big deal!

Reply:
Makes sense, I'll add the check! It's a subset, since we are ok with empty identities

Reply:
Actually just checked, it shouldn't be possible to pass an identity without a subject because empty lines are skipped over.